### PR TITLE
restore-dbadatabase working

### PR DESCRIPTION
**Pull request by potatoqualitee**
fix - got it working, @FriedrichWeinmann take a look

### CODE REVIEW

File: functions/Restore-DbaDatabase.ps1 [hunk 1]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
 						}
 					}
 				}
 				
-				Write-Verbose "$FunctionName - type = $($f.gettype())"
-				if ($f -is [string])
-				{
+				Write-Message -Level Verbose -Message "type = $($f.gettype())"
+				if ($f -is [string]) {
 					Write-Verbose "$FunctionName : Paths passed in"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Verbose must be gone

File: functions/Restore-DbaDatabase.ps1 [hunk 2]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
+						if ($XpDirTree) {
+							if ($p -match '\.\w{3}\Z') {
+								if (Test-DbaSqlPath -Path $p -SqlInstance $SqlInstance -SqlCredential $SqlCredential) {
+									$p = $p | Select-Object *, @{ Name = "FullName"; Expression = { $p } }
+									$backupFiles += $p
 								}
-								else
-								{
+								else {
 									Write-Warning "$FunctionName - $p cannot be accessed by $SqlInstance"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: No Warnings

File: functions/Restore-DbaDatabase.ps1 [hunk 3]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
+							catch {
+								if (Test-DbaSqlPath -Path $p -SqlInstance $SqlInstance -SqlCredential $SqlCredential) {
+									$p = $p | Select-Object *, @{ Name = "FullName"; Expression = { $p } }
+									$backupFiles += $p
 								}
-								else
-								{
+								else {
 									Write-Warning "$FunctionName - $p cannot be accessed by $SqlInstance"
 									continue
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Could use some flow control.
Stop-Function might be an idea.

File: functions/Restore-DbaDatabase.ps1 [hunk 4]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
 								}
 							}
 						}
-						elseif ($MaintenanceSolutionBackup)
-						{
+						elseif ($MaintenanceSolutionBackup) {
 							Write-Verbose "$FunctionName : Ola Style Folder"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Verbose

File: functions/Restore-DbaDatabase.ps1 [hunk 5]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
-							$BackupFiles += Get-OlaHRestoreFile -Path $p
+							$backupFiles += Get-OlaHRestoreFile -Path $p
 						}
-						else
-						{
+						else {
 							Write-Verbose "$FunctionName : Standard Directory"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Verbose

File: functions/Restore-DbaDatabase.ps1 [hunk 6]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
+							if ((($backupFiles.count) - $FileCheck) -eq 0) {
+								$backupFiles += Get-OlaHRestoreFile -Path $p
 							}
 						}
 					}
 				}
-				elseif (($f -is [System.IO.FileInfo]) -or ($f -is [System.Object] -and $f.FullName.Length -ne 0))
-				{
+				elseif (($f -is [System.IO.FileInfo]) -or ($f -is [System.Object] -and $f.FullName.Length -ne 0)) {
 					Write-Verbose "$FunctionName : Files passed in $($Path.count)"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Verbose

File: functions/Restore-DbaDatabase.ps1 [hunk 7]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
-									$BackupFiles += $FileTmp
+							else {
+								if (Test-DbaSqlPath -Path $FileTmp.FullName -SqlInstance $SqlInstance -SqlCredential $SqlCredential) {
+									$backupFiles += $FileTmp
 								}
-								else
-								{
-									Write-Warning "$FunctionName - $($FileTmp.FullName) cannot be access by $SqlInstance" 
+								else {
+									Write-Warning "$FunctionName - $($FileTmp.FullName) cannot be access by $SqlInstance"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Either `Write-Message` or `Stop-Function`

File: functions/Restore-DbaDatabase.ps1 [hunk 8]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
+						elseif (($FileTmp -is [System.Management.Automation.PSCustomObject])) #Dealing with Pipeline input 					
+{
+							Write-Message -Level Verbose -Message "Should be pipe input "
+							if ($FileTmp.PSobject.Properties.name -match "Server") {
 								#Most likely incoming from Get-DbaBackupHistory
-								if($Filetmp.Server -ne $SqlInstance -and $FileTmp.FullName -notlike '\\*')
-								{
+								if ($Filetmp.Server -ne $SqlInstance -and $FileTmp.FullName -notlike '\\*') {
 									Write-Warning "$FunctionName - Backups from a different server and on a local drive, can't access"
 									return
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Looks like a `Stop-Function`

File: functions/Restore-DbaDatabase.ps1 [hunk 9]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
+								Write-Message -Level Verbose -Message "it's folder"
+								ForEach ($ft in $Filetmp.FullName) {
+									Write-Message -Level Verbose -Message "Piped files Test-DbaSqlPath $($ft)"
+									if (Test-DbaSqlPath -Path $ft -SqlInstance $SqlInstance -SqlCredential $SqlCredential) {
+										$backupFiles += $ft
 									}
-									else
-									{
+									else {
 										Write-Warning "$FunctionName - $($ft) cannot be accessed by $SqlInstance"
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Warning

File: functions/Restore-DbaDatabase.ps1 [hunk 10]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
-		{
-			If (($null -ne $Server.Databases[$DatabaseName]) -and ($WithReplace -eq $false))
-			{
+	end {
+		if (Test-FunctionInterrupt) { return }
+		
+		if ($null -ne $DatabaseName) {
+			If (($null -ne $server.Databases[$DatabaseName]) -and ($WithReplace -eq $false)) {
 				Write-Warning "$FunctionName - $DatabaseName exists on Sql Instance $SqlInstance , must specify WithReplace to continue"
 				break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: `Stop-Function` case

File: functions/Restore-DbaDatabase.ps1 [hunk 11]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
-					if ((New-DbaSqlDirectory -Path $DestinationDataDirectory -SqlInstance $sqlinstance -SqlCredential $SqlCredential).Created -ne $true)
-					{
+		
+		if ($isLocal -eq $false) {
+			Write-Message -Level Verbose -Message "Remote server, checking folders"
+			if ($DestinationDataDirectory -ne '') {
+				if ((Test-DbaSqlPath -Path $DestinationDataDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential) -ne $true) {
+					if ((New-DbaSqlDirectory -Path $DestinationDataDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential).Created -ne $true) {
 						Write-Warning "$FunctionName - DestinationDataDirectory $DestinationDataDirectory does not exist, and could not be created on $SqlInstance"
 						break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: `Stop-Function` case

File: functions/Restore-DbaDatabase.ps1 [hunk 12]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
-			{
-				if ((Test-DbaSqlPath -Path $DestinationLogDirectory -SqlInstance $sqlinstance -SqlCredential $SqlCredential) -ne $true)
-				{
-					if((New-DbaSqlDirectory -Path $DestinationLogDirectory -SqlInstance $sqlinstance -SqlCredential $SqlCredential).Created -ne $true)
-					{
+			if ($DestinationLogDirectory -ne '') {
+				if ((Test-DbaSqlPath -Path $DestinationLogDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential) -ne $true) {
+					if ((New-DbaSqlDirectory -Path $DestinationLogDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential).Created -ne $true) {
 						Write-Warning "$FunctionName - DestinationLogDirectory $DestinationLogDirectory does not exist, and could not be created on $SqlInstance"
 						break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: `Stop-Function` case

File: functions/Restore-DbaDatabase.ps1 [hunk 13]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
 		#$AllFilteredFiles
 		#return
 		
-
-		if ($AllFilteredFiles.count -gt 1 -and $DatabaseName -ne '')
-		{
+		
+		if ($AllFilteredFiles.count -gt 1 -and $DatabaseName -ne '') {
 			Write-Warning "$FunctionName -  DatabaseName parameter and multiple database restores is not compatible "
 			break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: `Stop-Function` Case

File: functions/Restore-DbaDatabase.ps1 [hunk 14]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
 			
-			Write-Verbose "$FunctionName - Starting FileSet"
-			if (($FilteredFiles.DatabaseName | Group-Object | Measure-Object).count -gt 1)
-			{
+			
+			Write-Message -Level Verbose -Message "Starting FileSet"
+			if (($FilteredFiles.DatabaseName | Group-Object | Measure-Object).count -gt 1) {
 				$dbs = ($FilteredFiles | Select-Object -Property DatabaseName) -join (',')
 				Write-Warning "$FunctionName - We can only handle 1 Database at a time - $dbs"
 				break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: `Stop-Function` Case

File: functions/Restore-DbaDatabase.ps1 [hunk 15]
```diff
@@ -339,273 +304,209 @@ folder for those file types as defined on the target instance.
...
+				try {
+					$FilteredFiles | Restore-DBFromFilteredArray -SqlInstance $SqlInstance -DBName $databasename -SqlCredential $SqlCredential -RestoreTime $RestoreTime -DestinationDataDirectory $DestinationDataDirectory -DestinationLogDirectory $DestinationLogDirectory -NoRecovery:$NoRecovery -TrustDbBackupHistory:$TrustDbBackupHistory -ReplaceDatabase:$WithReplace -ScriptOnly:$OutputScriptOnly -FileStructure $FileMapping -VerifyOnly:$VerifyOnly -UseDestinationDefaultDirectories:$useDestinationDefaultDirectories -ReuseSourceFolderStructure:$ReuseSourceFolderStructure -DestinationFilePrefix $DestinationFilePrefix -MaxTransferSize $MaxTransferSize -BufferCount $BufferCount -BlockSize $BlockSize
 					$Completed = 'successfully'
 				}
-				catch
-				{
+				catch {
 					Write-Exception $_
 					$Completed = 'unsuccessfully'
 					return
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: `Stop-Function` Case. Finally should be removed (was bulk-renamed)

**Comment by potatoqualitee** [comment]
damn, i gotta sleep soon cuz of the presentation, can you fix? if not, ill just run this tomorrow. it works ;)

**Comment by potatoqualitee** [comment]
gross wait wtf
i pasted your new command right? why does this have old shit?

**Comment by potatoqualitee** [comment]
merging till asap!

**Comment by potatoqualitee** [comment]
wait, this is all fucked. im going back and just doing object and we'll figure it out later

**Comment by FriedrichWeinmann** [comment]
Do so, I'll see when/how I get to it.

**Comment by Stuart-Moore** [comment]
This is going to make merging the new stuff 'interesting'. Switching
messaging styles was scheduled for Thursday after I'd redone some of the
logic

What are you having issues with?

**Comment by potatoqualitee** [comment]
@Stuart-Moore  ahh it's Test-DbaLastBackup - Restore-DbaDatabase cant find trusthistory

**Comment by potatoqualitee** [comment]
Removing that switch made it work, it seems, lemme see if i restore the restore if it works, then ill just remove the trusthistory from test

**Comment by potatoqualitee** [comment]
hell yes, it's an issue with test-dbalastbackup, removing that trustbackuphistory param for restore in dev

**Comment by Stuart-Moore** [comment]
Yes. It's version dependant atm if it's the issue I think it is. Was going
to put an type check in to make it more reliable.

Those input validations are a nightmare as each type is different.
Hopefully with types it'll be simpler

**Pull request merged**